### ISSUE TITLE
Allow [FromQuery] to work with simple objects in Minimal API.

### ISSUE THREAD

**Issue by mrpmorris** [dotnet/aspnetcore#42438]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

Minimal API currently supports `[FromBody] SomeObject request` as a parameter, whereas `[FromQuery]` only supports simple types.

### Describe the solution you'd like

It would be less code and more intuitive to support objects with `[FromQuery]` too. For example

```
app.MapGet("/users/search", ([FromQuery] UserSearchQuery query) => Mediator.Send(query);
```

Just as we do for posts

```
app.MapPost("/users/search", ([FromBody] UserSearchQuery query) => Mediator.Send(query);
```

Note that having attributes decorating the properties of the query object, or requiring it to implement a static method to parse a query string would both be more work and also enforce the Contracts project have a dependency on ASP.NET, which is undesirable as the contract class should have no concept of how it came to exist.

So I propose the `[FromQuery]` approach being implemented in the ASP.NET level.

### Additional context

David Fowler = https://twitter.com/davidfowl/status/1541088042169008130

**Comment by davidfowl** [comment]
@brunolins16 FWIW, I think this would look exactly like `[AsParameters]`. No nesting of these within the object graph and only a single binding source. I wonder if that logic can be generalized? Also if there was a name specified on this attribute it would fail if there was no custom TryParse method.

**Comment by mrpmorris** [comment]
I don't think it should require a TryParse method. The idea is that the object itself is completely agnostic as to how it is being used (Http, ServiceBus, etc).

Just a one-to-one mapping from each value named in the QueryString to property in the class.

**Comment by brunolins16** [comment]
> I wonder if that logic can be generalized?

Do you mean, we should have same behavior for other sources, eg. `FromHeader`?

> I think this would look exactly like `[AsParameters]`

Just to make sure I understood. Will both examples have the same behavior?

``` C#
public struct UserSearchQuery
{
   // Attributes are not required but added just to make the example clearer.
   [FromQuery]
   public string FirstName { get; set;}
   [FromQuery]
   public string LastName { get; set;}
}
app.MapGet("/users/search", ([AsParameters] UserSearchQuery query) => Mediator.Send(query);
``` 

and

``` C#
public struct UserSearchQuery
{
   public string FirstName { get; set;}
   public string LastName { get; set;}
}
app.MapGet("/users/search", ([FromQuery] UserSearchQuery query) => Mediator.Send(query);
```

**Comment by brunolins16** [comment]
A different between `AsParameters` and this new `FromQuery` will be how we handle `properties` attributes because when using `FromQuery` we should ignore or throw if a binding source attribute is present, right?

**Comment by davidfowl** [comment]
> Just to make sure I understood. Will both examples have the same behavior?

Yes, that's right.

> A difference between AsParameters and this new FromQuery will be how we handle properties attributes because when using FromQuery we should ignore or throw if a binding source attribute is present, right?

Throw.

**Comment by diegosasw** [comment]
I've found the conversation on Twitter.

I had also noticed in .NET 7 the minimal API `[FromQuery]` behaves differently than when using the endpoint in a controller extending `ControllerBase`, as the [FromQuery] in minimal API, which seems to require adding additional things to the filter model.
> No public static bool AgentTenantedFilter.TryParse(string, out AgentTenantedFilter) method found for filter.

I suppose this is the same issue you were suggesting a new feature for?

`[AsParameters]` is not an equivalent. In my case, since my filter extends some functionality, Swagger/OpenAPI is marking the inherited filter properties as required.

TL;DR; I just wanted to confirm that currently in .NET 7 minimal API `[FromQuery]` behaves differently by design, which would explain why I can't move my controller
```
[Route("agents")]
public class AgentQueryApi : QueryControllerBase // which is a ControllerBase
{
	private readonly FilteredQueryService<AgentDocument, AgentTenantedFilter> _filteredQueryService;

	public AgentQueryApi(FilteredQueryService<AgentDocument, AgentTenantedFilter> filteredQueryService)
		=> _filteredQueryService = filteredQueryService;

	[HttpGet]
	public async Task<PagedResult<AgentDocument>> GetAll(
		[FromQuery] AgentTenantedFilter agentTenantedFilter,
		CancellationToken cancellationToken)
	{
		var result = await _filteredQueryService.Get(agentTenantedFilter, cancellationToken);
		return result;
	}
}
```

to a minimal API
```
app.MapGet("agents",
    async (FilteredQueryService<AgentDocument, AgentTenantedFilter> filteredQueryService,
    [FromQuery] AgentTenantedFilter agentTenantedFilter,
        CancellationToken cancellationToken) =>
    {
        // It won't work anyway, as Swagger complains
    })
```

Is it because the `ControllerBase` is adding some binding behavior? Will Minimal API be able to ever work in the same way so that migrating from `ControllerBase` to minimal API is trivial?

Sorry if my question is not really related to the original motivation.

**Comment by mitchdenny** [comment]
@davidfowl I'm wondering what the value is of `[FromQuery]` being used to replicate the behavior of `[AsParameters]`? Doesn't that just make things more confusing?

**Comment by Seabizkit** [comment]
can not do 
```
app.MapPost("/ReadCsv", async (IFormFile file, [FromQuery] OptionsRequest options) =>
{
    return Results.Ok($"");
}).DisableAntiforgery()
.WithName("ReadCsv");
```

can not do 
```
app.MapPost("/ReadCsv", async (IFormFile file, [FromQuery] CsvRequest options) =>
{
    return Results.Ok($"");
}).DisableAntiforgery()
.WithName("ReadCsv");

app.Run();

public class CsvRequest 
{
    [FromForm] public  IFormFile File { get; set; }

    [FromQuery] public OptionsRequest Options { get; set; }
}
```


I would really like to see this supported.

im trying to convert controller which looks like 

```
[HttpPost("ReadCsv")]
public async Task<IActionResult> ReadCsv(IFormFile file, [FromQuery] OptionsRequest request, [FromQuery] CsvConfig csvConfig = null)
{
}
```

i would really love to see better support for model binding..., at least on par with controllers
it would make moving to minimal API acually possiable.

**Comment by davidfowl** [comment]
Why not use AsParameters?

**Comment by Seabizkit** [comment]
thanks, maybe its just not obvious or well confusing with all the attrubutes.

this does work, which was not my first guess.. 
```
app.MapPost("/api/v1/TextReader/ReadCsv", async ( 
    ITextReaderService textReaderService,
    IFormFile file,
    [AsParameters] OptionsRequest request,
    [AsParameters] CsvConfig config = null) =>
{
```

is AsParameters the new FromQuery? 
to me, the below should work..
```
app.MapPost("/api/v1/TextReader/ReadCsv", async ( 
    ITextReaderService textReaderService,
    IFormFile file,
    [FromQuery] OptionsRequest request,
    [FromQuery] CsvConfig config = null) =>
{
```

**Comment by davidfowl** [comment]
https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/parameter-binding?view=aspnetcore-9.0#parameter-binding-for-argument-lists-with-asparameters

**Comment by mrpmorris** [comment]
Thanks for letting me know!
